ErrorWithPos.GetPosition() panics if Pos is nil

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1509fa6]

goroutine 79 [running]:
github.com/jhump/protoreflect/desc/protoparse.ErrorWithSourcePos.GetPosition(...)
	/go/pkg/mod/github.com/jhump/protoreflect@v1.6.1/desc/protoparse/errors.go:104
```

The issue is that `GetPosition()` deferences the `Pos` pointer, and there's no validation that this pointer is always set. Will investigate the code now.

--- a/desc/protoparse/errors.go
+++ b/desc/protoparse/errors.go
@@ -121,3 +121,16 @@ var _ ErrorWithPos = ErrorWithSourcePos{}
 func errorWithPos(pos *SourcePos, format string, args ...interface{}) ErrorWithPos {
 	return ErrorWithSourcePos{Pos: pos, Underlying: fmt.Errorf(format, args...)}
 }
+
+type errorWithFilename struct {
+	underlying error
+	filename   string
+}
+
+func (e errorWithFilename) Error() string {
+	return fmt.Sprintf("%s: %v", e.filename, e.underlying)
+}
+
+func (e errorWithFilename) Unwrap() error {
+	return e.underlying
+}

--- a/desc/protoparse/parser.go
+++ b/desc/protoparse/parser.go
@@ -449,7 +449,8 @@ func parseProtoFile(acc FileAccessor, filename string, importLoc *SourcePos, err
 	} else {
 		if !strings.Contains(err.Error(), filename) {
 			// an error message that doesn't indicate the file is awful!
-			err = fmt.Errorf("%s: %w", filename, err)
+			// this cannot be %w as this is not compatible with go <= 1.13
+			err = fmt.Errorf("%s: %v", filename, err)
 		}
 		// The top-level loop in parseProtoFiles calls this with nil for the top-level files
 		// importLoc is only for imports, otherwise we do not want to return a ErrorWithSourcePos

--- a/desc/protoparse/parser.go
+++ b/desc/protoparse/parser.go
@@ -408,14 +408,14 @@ func fixupFilenames(protos map[string]*parseResult) map[string]*parseResult {
 
 func parseProtoFiles(acc FileAccessor, filenames []string, errs *errorHandler, recursive, validate bool, parsed *parseResults, lookupImport func(string) (*desc.FileDescriptor, error)) {
 	for _, name := range filenames {
-		parseProtoFile(acc, name, nil, errs, recursive, validate, parsed, lookupImport)
+		parseProtoFile(acc, name, &SourcePos{Filename: name}, errs, recursive, validate, parsed, lookupImport)
 		if errs.err != nil {
 			return
 		}
 	}
 }
 
-func parseProtoFile(acc FileAccessor, filename string, importLoc *SourcePos, errs *errorHandler, recursive, validate bool, parsed *parseResults, lookupImport func(string) (*desc.FileDescriptor, error)) {
+func parseProtoFile(acc FileAccessor, filename string, loc *SourcePos, errs *errorHandler, recursive, validate bool, parsed *parseResults, lookupImport func(string) (*desc.FileDescriptor, error)) {
 	if parsed.has(filename) {
 		return
 	}
@@ -447,13 +447,9 @@ func parseProtoFile(acc FileAccessor, filename string, importLoc *SourcePos, err
 		//  parsers are trying to access it concurrently)
 		result = &parseResult{fd: proto.Clone(d).(*dpb.FileDescriptorProto)}
 	} else {
-		if !strings.Contains(err.Error(), filename) {
-			// an error message that doesn't indicate the file is awful!
-			err = fmt.Errorf("%s: %v", filename, err)
-		}
 		// associate the error with the import line
 		err = ErrorWithSourcePos{
-			Pos:        importLoc,
+			Pos:        loc,
 			Underlying: err,
 		}
 		_ = errs.handleError(err)

--- a/desc/protoparse/parser.go
+++ b/desc/protoparse/parser.go
@@ -450,7 +450,10 @@ func parseProtoFile(acc FileAccessor, filename string, importLoc *SourcePos, err
 		if !strings.Contains(err.Error(), filename) {
 			// an error message that doesn't indicate the file is awful!
 			// this cannot be %w as this is not compatible with go <= 1.13
-			err = fmt.Errorf("%s: %v", filename, err)
+			err = errorWithFilename{
+				underlying: err,
+				filename:   filename,
+			}
 		}
 		// The top-level loop in parseProtoFiles calls this with nil for the top-level files
 		// importLoc is only for imports, otherwise we do not want to return a ErrorWithSourcePos

--- a/desc/protoparse/parser.go
+++ b/desc/protoparse/parser.go
@@ -408,14 +408,14 @@ func fixupFilenames(protos map[string]*parseResult) map[string]*parseResult {
 
 func parseProtoFiles(acc FileAccessor, filenames []string, errs *errorHandler, recursive, validate bool, parsed *parseResults, lookupImport func(string) (*desc.FileDescriptor, error)) {
 	for _, name := range filenames {
-		parseProtoFile(acc, name, &SourcePos{Filename: name}, errs, recursive, validate, parsed, lookupImport)
+		parseProtoFile(acc, name, nil, errs, recursive, validate, parsed, lookupImport)
 		if errs.err != nil {
 			return
 		}
 	}
 }
 
-func parseProtoFile(acc FileAccessor, filename string, loc *SourcePos, errs *errorHandler, recursive, validate bool, parsed *parseResults, lookupImport func(string) (*desc.FileDescriptor, error)) {
+func parseProtoFile(acc FileAccessor, filename string, importLoc *SourcePos, errs *errorHandler, recursive, validate bool, parsed *parseResults, lookupImport func(string) (*desc.FileDescriptor, error)) {
 	if parsed.has(filename) {
 		return
 	}
@@ -447,10 +447,19 @@ func parseProtoFile(acc FileAccessor, filename string, loc *SourcePos, errs *err
 		//  parsers are trying to access it concurrently)
 		result = &parseResult{fd: proto.Clone(d).(*dpb.FileDescriptorProto)}
 	} else {
-		// associate the error with the import line
-		err = ErrorWithSourcePos{
-			Pos:        loc,
-			Underlying: err,
+		if !strings.Contains(err.Error(), filename) {
+			// an error message that doesn't indicate the file is awful!
+			err = fmt.Errorf("%s: %w", filename, err)
+		}
+		// The top-level loop in parseProtoFiles calls this with nil for the top-level files
+		// importLoc is only for imports, otherwise we do not want to return a ErrorWithSourcePos
+		// ErrorWithSourcePos should always have a non-nil SourcePos
+		if importLoc != nil {
+			// associate the error with the import line
+			err = ErrorWithSourcePos{
+				Pos:        importLoc,
+				Underlying: err,
+			}
 		}
 		_ = errs.handleError(err)
 		return

